docs/proposed/hs_reassembly.md

```diff
@@ -0,0 +1,410 @@
+
+```
+ +-------------------------+ +--------------------------+
+ |      TLS record m       | |      TLS record m+1      |
+ |+----------+ +----------+| |+------------++----------+|
+ ||   HS k   | |HS k+1 ...|| || ... HS k+1 ||HS k+2 ...||
+ |+----------+ +----------+| |+------------++----------+|
+ +-------------------------+ +--------------------------+
+                             |
+                             +------+
+                                    v
+                           +----------------+
+                           |  Accumulator   |
+                           +----------------+
+                                    |
+                                    +----------+
+                                               v
+                               +----------+ +-----------------+
+                               |   HS k   | |     HS k+1      |
+                               +----------+ +-----------------+
+
+```
+
+Conceptually, the `mbedtls_reader` is a FIFO buffer, that offers the following API:
+
+- `mbedtls_reader_feed()`    - appends data to the "tail" of the FIFO.
+- `mbedtls_reader_get()`     - provides access to the "head" of the FIFO.
+- `mbedtls_reader_commit()`  - informs the "FIFO" that the data that have been previously
+                               requested via `mbedtls_reader_get()` have been processed.
+- `mbedtls_reader_reclaim()` - cleans up the space that's occupied by the already committed
+                               messages.
+
+## Data structures
+
+Reassembly must be enabled by uncommenting `MBEDTLS_SSL_HS_REASSEMBLY`
+in the `include/mbedtls/config.h` header file, and also enabled in the runtime
+by setting a non-zero max message size.
+
+
+Suggested configuration preprocessor constructs:
+
+```
+/** Enable handshake reassembly. */
+#define MBEDTLS_SSL_HS_REASSEMBLY
+
+/** Hard limit for the maximal message size. */
+#define MBEDTLS_SSL_HS_REASSEMBLY_MAX_MESSAGE_SIZE
+
+/** The operation can not be performed because of the maximal message size limitation. */
+#define MBEDTLS_ERR_SSL_HS_REASSEMBLY_MAX_MESSAGE_SIZE_EXCEEDED
+```
+
+Suggested API to enable reassembly in runtime:
+
+```
+struct mbedtls_ssl_config {
+
+...
+
+#if defined(MBEDTLS_SSL_HS_REASSEMBLY)
+    /** Maximal supported message size. Value 0 indicates that the reassembly is disabled.*/
+    size_t hs_reassembly_max_message_size;
+#endif /* MBEDTLS_SSL_HS_REASSEMBLY */
+
+...
+
+};
+
+/**
+ * Set max size of reassembled message.
+ *
+ * \param conf mbedtls config.
+ * \param max_message_size - maximal allowed size for the reassembled message.
+ *                           Must not exceed MBEDTLS_SSL_HS_REASSEMBLY_MAX_MESSAGE_SIZE.
+ * \return 0 on success, or MBEDTLS_ERR_SSL_HS_REASSEMBLY_MAX_MESSAGE_SIZE_EXCEEDED on failure.
+ */
+int mbedtls_ssl_conf_hs_reassembly_set_max_message_size( mbedtls_ssl_config *conf, size_t max_message_size );
+```
+
+
+Suggested implementation in `mbedtls_ssl_context`:
```

Comment:
This should be a separate section `## Implementation details` or similar.

Reply:
Addressed in the design doc.